pkg/kv/kvserver/kvflowcontrol/kvflowconnectedstream/stream.go

```diff
@@ -0,0 +1,1339 @@
+	// Replica.mu), and so this map can be read with either of these mutexes
+	// (and we will read it with Replica.mu).
+	//
+	// Unlike v1, where this destroyed the connected-stream data-structure for
+	// this replica, we could just return the inflight send tokens and keep the
+	// rest of the state (send-queue and stats, and eval token deductions) since
+	// it may still be in StateReplicate. Though if this is actually down and we
+	// stay in StateReplicate, we are penalizing elastic work which will still
+	// pace at rate of slowest replica (and the rate of this replica will be 0).
+	// Actually, we should keep the connected-stream for a couple of ticks. This
+	// will also ensure that this stream can participate in the quorum, and we
+	// don't force-flush. But this will also be building up a send-queue, so we
+	// may have a quorum of connected-streams with tokens, but we don't have a
+	// quorum with an empty send-queue -- figure out a reasonable heuristic
+	// (perhaps wait to force-flush until the non-empty send-queue has existed
+	// for more than a tick).
+	//
+	// TODO: this is what motivates having a mutex in RangeController.
+	TransportDisconnected(replica roachpb.ReplicaID)
+	// Close the controller, since no longer the leader. Can be called concurrently
+	// with other methods like WaitForEval. WaitForEval should unblock and return
+	// without an error.
+	Close()
+}
+
+// RaftEvent is an abstraction around raft.Ready, constructed in
+// handleRaftReadyRaftMuLocked.
+type RaftEvent interface {
+	// Ready can return nil if there is no Ready.
+	Ready() Ready
+}
+
+// Ready interface is an abstraction around raft.Ready, for the pull mode in
+// which replication AC will operate. In this pull mode, the Raft leader will
+// produce a Ready that has the new entries it knows about, but MsgApps will
+// only be produced for a follower for (a) entries in (Match, Next), i.e.,
+// entries that are already in-flight for that follower, and have been nacked,
+// (b) empty MsgApps to ping the follower. Next is only advanced via pull,
+// which is serviced via RaftInterface. In a sense, there is a NextUpperBound
+// maintained by Raft, that never regresses in StateReplicate -- it advances
+// whenever RaftInterface.MakeMsgApp is called. And everything in (Match,
+// NextUpperBound) is the responsibility of Raft to retry. The existing notion
+// of Next is protocol state internal to raft, regarding what to retry
+// sending. NextUpperBound can regress if the replica transitions out of
+// StateReplicate, and back into StateReplicate.
+//
+// The Ready described here is only the subset that is needed by replication
+// AC -- heartbeats, appending to the local log etc. are not relevant.
+//
+// Ready must be called on every tick/ready of Raft since we cannot tolerate
+// state transitions from
+//
+// StateReplicate => {state in all states: state != StateReplicate} =>
+// StateReplicate
```

Comment:
Are you saying that we can't tolerate an ABA problem in which we go out and back into `StateReplicate` and can't tell if the node has always been in this state or not?